packages/xchain-haven/src/client.ts

```diff
@@ -0,0 +1,315 @@
+        baseAmount(havenTx.total_sent[havenTx.from_asset_type], 12).gt('0') &&
+        (params?.asset === undefined || (params?.asset !== undefined && havenTx.from_asset_type == params.asset))
+
+      const isIn: boolean =
+        baseAmount(havenTx.total_received[havenTx.to_asset_type], 12).gt('0') &&
+        (params?.asset === undefined || (params?.asset !== undefined && havenTx.to_asset_type == params.asset))
+
+      const from: Array<TxFrom> = isOut
+        ? [
+            {
+              amount: baseAmount(havenTx.fromAmount, 12),
+              asset: getAssetByTicker(havenTx.from_asset_type),
+              from: havenTx.hash,
+            },
+          ]
+        : []
+      const to: Array<TxTo> = isIn
+        ? [
+            {
+              amount: baseAmount(havenTx.toAmount, 12),
+              asset: getAssetByTicker(havenTx.to_asset_type),
+              to: havenTx.hash,
+            },
+          ]
+        : []
+
+      const tx: Tx = {
+        hash: havenTx.hash,
+        date: new Date(havenTx.timestamp),
+        type: havenTx.from_asset_type === havenTx.to_asset_type ? TxType.Transfer : TxType.Unknown,
+        from,
+        to,
+        asset: isOut ? getAssetByTicker(havenTx.from_asset_type) : getAssetByTicker(havenTx.to_asset_type),
+      }
+
+      return tx
+    })
+
+    const txPages: TxsPage = {
+      txs,
+      total: txs.length,
+    }
+    return txPages
+  }
+  async getTransactionData(txId: string, _assetAddress?: string): Promise<Tx> {
+    const havenTx = await this.havenSDK.getTx(txId)
+
+    const isOut: boolean = baseAmount(havenTx.total_sent[havenTx.from_asset_type], 12).gt('0')
+    const isIn: boolean = baseAmount(havenTx.total_received[havenTx.to_asset_type], 12).gt('0')
+
+    const from: Array<TxFrom> = isOut
+      ? [
+          {
+            amount: baseAmount(havenTx.fromAmount, 12),
+            asset: getAssetByTicker(havenTx.from_asset_type),
+            from: havenTx.hash,
+          },
+        ]
+      : []
+    const to: Array<TxTo> = isIn
+      ? [
+          {
+            amount: baseAmount(havenTx.toAmount, 12),
+            asset: getAssetByTicker(havenTx.to_asset_type),
+            to: havenTx.hash,
+          },
+        ]
+      : []
+
+    const asset: Asset = isOut ? getAssetByTicker(havenTx.from_asset_type) : getAssetByTicker(havenTx.to_asset_type)
+
+    const tx: Tx = {
+      hash: havenTx.hash,
+      date: new Date(havenTx.timestamp),
+      type: isIn && isOut ? TxType.Unknown : TxType.Transfer,
+      from,
+      to,
+      asset,
+    }
+    return tx
+  }
+
+  async transfer(params: TxParams): Promise<TxHash> {
+    const { amount, asset, recipient, memo } = params
+    if (asset === undefined) throw new Error('please specify asset it in Client.transfer() for Haven')
+    const amountString = amount.amount().toString()
+    let txHash
+    try {
+      txHash = await this.havenSDK.transfer(amountString, asset.ticker as HavenTicker, recipient, memo)
```

Comment:
(Similar to a previous comment) Try to use a type guard instead of `as HavenTicker`